tephra-core/src/main/java/co/cask/tephra/TransactionContext.java

```diff
@@ -83,11 +100,53 @@ public void finish() throws TransactionFailureException {
     currentTx = null;
   }
 
+  /**
+   * Aborts the given transaction, and rolls back all data set changes. If rollback fails,
+   * the transaction is invalidated. If an exception is caught during rollback, the exception
+   * is rethrown wrapped in a TransactionFailureException, after all remaining TransactionAwares have
+   * completed rollback.
+   *
+   * @throws TransactionFailureException for any exception that is encountered.
+   */
   public void abort() throws TransactionFailureException {
     abort(null);
   }
 
   /**
+   * Checkpoints the current transaction by flushing any pending writes for the registered {@link TransactionAware}
+   * instances, and obtaining a new current write pointer for the transaction.  By performing a checkpoint,
+   * the client can ensure that all previous writes were flushed and are visible.  By default, the current write
+   * pointer for the transaction is also visible.  The current write pointer can be excluded from read
+   * operations by setting an attribute with the key {@link TxConstants#TX_EXCLUDE_CURRENT_WRITE} on the
+   * {@code Get} or {@code Scan} operation.  After the checkpoint operation is performed, the updated
+   * {@link Transaction} instance will be passed to {@link TransactionAware#startTx(Transaction)} for each
+   * registered {@code TransactionAware} instance.
+   *
+   * @throws TransactionFailureException if an error occurs while performing the checkpoint
+   */
+  public void checkpoint() throws TransactionFailureException {
+    Preconditions.checkState(currentTx != null, "Cannot checkpoint tx that has not been started");
+    persist();
+    try {
+      currentTx = txClient.checkpoint(currentTx);
+      // update the current transaction with all TransactionAwares
+      for (TransactionAware txAware : txAwares) {
+        txAware.startTx(currentTx);
```

Comment:
would it make sense to get the change set from all tx-awares here? If we do that, then tx-awares don't even need to know whether they are part of a check-pointed transaction or not.